src/advanced.sml

```diff
@@ -0,0 +1,416 @@
+  fun cluedSuit (is : info list) : suit option =
+    case List.mapPartial (fn i => case i of IsSuit su => SOME su | _ => NONE) is of
+         [] => NONE
+       | x::xs => if List.exists (fn su => su <> x) xs then SOME Rainbow else SOME x
+
+  (* Returns the set of possible suits of a card. *)
+  fun possibleSuits (is : info list) : SSet.set =
+    case is of
+         (IsSuit su)::xs => SSet.intersection (possibleSuits xs) (rainbowAnd su)
+       | (NotSuit su)::xs => SSet.difference (possibleSuits xs) (rainbowAnd su)
+       | _::xs => possibleSuits xs
+       | [] => suits
+
+  exception Unreachable
+  fun possibleSuitsAndRanks (is : info list) : SSet.set * RSet.set =
+    case is of
+         [] => (suits, ranks)
+       | (IsRank r)::xs => (possibleSuits xs, RSet.singleton r)
+       | i::xs => let val (sus,rs) = possibleSuitsAndRanks xs
+                  in case i of
+                          IsSuit su => (SSet.intersection sus (rainbowAnd su), rs)
+                        | NotSuit su => (SSet.difference sus (rainbowAnd su), rs)
+                        | NotRank r => (sus, RSet.remove rs r)
+                        | _ => raise Unreachable
+
+                  end
+
+
+  (* Is this card guaranteed to be useless? *)
+  fun isUseless' (s : state) (is : info list) : bool =
+  let
+    val maxRank = RSet.foldl Int.max 0 (possibleRanks is)
+    val allSuits = SSet.toList (possibleSuits is)
+  in
+    List.all (fn su => maxRank <= SD.lookup (#inPlay s) su) allSuits
+  end
+
+  fun ourOldestUseless (s : state) (xs : info list list) : int option =
+    Util.revFindIndex (isUseless' s) xs
+
+  (* Returns whether a card is definitely not playable *)
+  fun isNotPlayable (s : state) (is : info list) : bool =
+  let
+    val sus = SSet.toList (possibleSuits is)
+    val rs = possibleRanks is
+  in
+    List.all (fn su => not (RSet.member rs (SD.lookup (#inPlay s) su + 1))) sus
+  end
+
+  (* Returns whether a card is definitely playable *)
+  fun isDefinitelyPlayable (s : state) (is : info list) : bool =
+  let
+    val sus = SSet.toList (possibleSuits is)
+    val rs = RSet.toList (possibleRanks is)
+  in
+    length rs = 1 andalso List.all (fn su => hd rs = SD.lookup (#inPlay s) su + 1) sus
+  end
+
+  datatype fuzzyBool = True | False | Maybe
+
+  fun isPlayableFuzzy (s : state) (is : info list) : fuzzyBool =
+  let
+    val (sus', rs') = possibleSuitsAndRanks is
+    val sus = SSet.toList sus'
+    val rs = RSet.toList rs'
+  in
+    if length rs = 1 andalso List.all (fn su => hd rs = SD.lookup (#inPlay s) su + 1) sus
+    then True
+    else if List.all (fn su => not (RSet.member rs' (SD.lookup (#inPlay s) su + 1))) sus
+    then False
+    else
+        Maybe
+  end
+
+  fun FuzzyBoolToBool (k : fuzzyBool) (default : bool) : bool =
+  case k of
+       True => true
+     | False => false
+     | Maybe => default
+
+  (* check whether a card in your own hand is definitely not vital *)
+  fun isNotVital (s : state) (is : info list) : bool =
```

Comment:
This seems like it might be expensive...it might be better to implement this in a way that isn't as bad when `possibleSuits` or `possibleRanks` are the full sets.